[Team-09][BE][테리&벅픽] 2주차 PR 1번째 - 로그인 API 구현 마무리, Issue 저장 API 구현

mybloom opened the pull request:
안녕하세요 `프레디`!
`테리, 벅픽` 입니다. 2주차 1번째 PR 입니다.
세심한 리뷰 항상 감사드립니다!! 😀


리뷰 받고 싶은 부분은 코멘트로 추가하였습니다!
감사합니다!


### ERD 및 API docs

<details>
<summary> ERD </summary>
    
![image](https://user-images.githubusercontent.com/91416897/174514679-dcae6a60-c8a8-4284-9160-c3575af76cfe.png)
    
</details>

<details>
<summary> API docs </summary>
    
https://documenter.getpostman.com/view/20452796/UzBiNobv
    
</details>

### 작업 내용


<summary> ✅  Issue 생성 API 작성 </summary>

- 유의해서 개발한 부분 : 검증 로직
    - Issue 생성을 위해 연관된  마일스톤, 레이블, 할당자 를  request Body에서 가져와 DB에 저장합니다. 
    - DB 저장 전에 연관된 마일스톤, 레이블, 할당자 가 실제 존재하며,  이슈 생성자가 생성한 정보인지 확인하는 검증 로직을 작성했습니다. 해당 검증로직은 고민하다가 `IssueValidateService` 에 작성하였습니다.
- 🧡 `리뷰 받고 싶은 부분`    
    - 로직을 짜는 것이 익숙하지 않고 어렵습니다.  소소하더라도 `로직 구현`에 대해 리뷰해 주신다면 많은 도움이 될 것 같습니다. 





<summary> ✅ Login API 마무리 </summary>
    
- 로그인 관련하여 전체 흐름도는 다음과 같습니다.
    ![image](https://user-images.githubusercontent.com/91416897/174982228-48e8dd2c-992d-4c01-bfea-5fac8cc95cfb.png)
    
- Spring Filter를 통해 검증이 필요한 URL에 대하여 JWT access token을 검증하는 로직을 추가하였습니다.    

- RedisDB를 통한 refresh 토큰 저장 기능을 추가하였습니다.

- 만약 프론트에서 만료된 access token으로 접근시에, 예외를 응답합니다. 아직 예외처리 핸들러는 추가하지 않아, 스프링 기본 설정 예외가 반환됩니다. 예외처리 핸들러는 추후 추가할 예정입니다.
    
- 프론트에서는 access token과 refresh token을 요청 바디에 넣어 토큰 갱신 요청을 보냅니다. 서버에서는 RedisDB에 저장되어 있는 refresh token을 바탕으로 검증을 하고, 새롭게 생성한 access token을 반환합니다.
    - refresh token의 경우 RedisDB에 저장될 때, 토큰의 만료시간만큼을 DB상의 만료시간으로 설정합니다.

mybloom committed:
sql.docs: member테이블에 user_id 유니크 제약조건 추가, email컬럼 추가  (bugpigg/issue-tracker#17)

- 서드파티 id 부분을 일반 계정 이메일과 분류

mybloom committed:
feat: 커스텀익셉션 생성  (bugpigg/issue-tracker#17)

- 상위 : BusinessException, RuntimeException 상속
- 중간 : InvalidException 유효하지 않은 값에 대한 익셉션 설계
- 하위 : 실제 도메인에서 사용하는 익셉션

mybloom committed:
feat: IssueServiceDto 생성 - Service와 엔티티 연결 Dto  (bugpigg/issue-tracker#17)

mybloom committed:
feat: Issuer관련 연관 엔티티에 객체 생성 정적팩토리 메서드 추가   (bugpigg/issue-tracker#17)

mybloom committed:
fix: MemberType enum 대문자로 변경  (bugpigg/issue-tracker#17)

- 자바 컨벤션

bugpigg committed:
[BE] 인증 관련 API 구현 2 (#23)

* build: RedisDB 관련 의존성 추가

* refactor: jwt configuration 클래스 명 수정

- JwtConfig -> JwtConfiguration 으로 클래스 명 수정

* feat: Redis 관련 설정 및 서비스 클래스 추가

- RedisConfiguration 에서는 Redis 관련 설정 상수들을 로드하고 redisTemplate을 스프링 빈으로 등록합니다.
- RedisService 에서는 RedisDB 관련 트랜잭션 연산을 수행하는 클래스입니다.

* feat: 서블릿 필터를 활용한 JWT 토큰 검증 기능 추가(bugpigg/issue-tracker#19)

- LoginValidateFilter 에서는 로그인이 필요한 요청에 대하여, JWT 토큰 기반 검증을 수행합니다. 만약 헤더에 토큰 값이 없거나 유효하지 않은 토큰인 경우 예외가 발생합니다.
- FilterConfiguration 에서는 LoginValidateFilter 를 스프링 빈으로 등록하고 검증을 위한 JwtTokenProvider 객체를 주입해줍니다.

* feat: 로그인 과정의 refreshToken을 RedisDB에 저장하는 기능 추가(bugpigg/issue-tracker#19)

* fix: JWT 토큰에 멤버의 DB상 인덱스 값이 아닌 아이디를 페이로드에 포함하게 수정

* feat: userId 기반으로 Member 엔티티의 id를 반환하는 함수 추가

- MemberService 클래스에 userId 기반으로 Member 엔티티의 id를 반환하는 함수를 추가하였습니다.
- LoginValidateFilter 클래스에서 MemberService의 findIdFromUserId 함수를 호출 가능하게 하기 위해, FilterConfiguration 클래스에서 LoginValidateFilter 객체를 스프링 빈으로 등록할때, MemberService 를 주입 받도록 수정하였습니다.

* feat: LoginValidateFilter 에 토큰 기반으로 멤버 ID 추출 및 응답 헤더에 추가 기능 구현

- LoginValidateFilter 에서 요청 헤더의 JWT 토큰을 기반으로 userId를 추출합니다. 그리고 MemberService 클래스의 findIdFromUserId() 함수를 통해 멤버 id를 구하고 이를 응답 헤더에 추가합니다. 이는 로그인이 요구되는 각 요청들에 대하여, 어떤 멤버가 보낸 요청인지 비즈니스 로직에서 식별할 필요성이 있기에 추가하였습니다. 필터에서 공통적으로 처리해주어 각 비즈니스 로직에서 중복될 수 있는 코드를 최소화 하였습니다.
- Request 에 추가하는게 적절하다고 판단하나, 구현상 어려워 Response 에 추가하였습니다.

* Fix: 토큰 만료 여부 예외 발생 추가

- 토큰이 만료되었을 경우 예외를 발생하게 하여, 프론트에서 재발급 URL로 다시 요청하도록 구현할 계획입니다. 현재는 우선 예외만 발생하게 하였습니다. 추후에 예외처리 핸들러를 추가할 계획입니다.
- 그리고 토큰 만료시, 재발급 URL은 로그인 검증 필터 로직을 거치지 않게 하기 위해 LoginValidateFilter의 VALID_FREE_URLS 을 수정하였습니다.

* feat: 토큰 만료시, access Token 재발급 기능 추가

- JWT access token 만료시에, access token을 재발급 하는 기능을 추가하였습니다.

* Fix: RedisDB 속 리프레시 토큰의 만료 시간 설정

mybloom committed:
feat: MilestoneId 유효성 체크 서비스 로직, Repository작성(bugpigg/issue-tracker#17)

mybloom committed:
feat: MemberId 유효성 체크 서비스 로직, Repository작성(bugpigg/issue-tracker#17)

mybloom committed:
feat: Member 엔티티에 of 정적팩토리 메서드 추가 (bugpigg/issue-tracker#17)

mybloom committed:
feat: LabelId 유효성 로직 , Repository 생성 (bugpigg/issue-tracker#17)

mybloom committed:
feat: IssueLabelRepository 생성 (bugpigg/issue-tracker#17)

- validation 체크용
- IssueLabel 저장용

mybloom committed:
feat: IssueRepository 에 findByMilestone() 추가(bugpigg/issue-tracker#17)

mybloom committed:
test: QueryMethod 테스트 코드 추가 (bugpigg/issue-tracker#17)

mybloom committed:
feat: Issue 저장,수정 시 입력 파라미터값 유효성 체크용 Service 생성 (bugpigg/issue-tracker#17)

mybloom committed:
feat: Issue 생성로직 Service에 구현  (bugpigg/issue-tracker#17)

mybloom committed:
docs: table 생성문 변경된 것 스크립트에 반영(bugpigg/issue-tracker#17)

mybloom committed:
feat: 연관관계 편의메서드 추가 (bugpigg/issue-tracker#17)

mybloom committed:
fix: 2차리뷰 boolean 리턴값 처리 (bugpigg/issue-tracker#17)

mybloom committed:
chore: 불필요한 주석 삭제 (bugpigg/issue-tracker#17)

mybloom committed:
Merge branch 'BE-feat-17' into BE-3

mybloom commented on BE/src/main/java/com/team09/issue_tracker/issue/domain/Issue.java:
```diff
@@ -51,32 +53,35 @@ public class Issue extends BaseTimeEntity {
 	@OneToMany(mappedBy = "issue", fetch = FetchType.LAZY)
 	private List<IssueLabel> issueLabels = new ArrayList<>();
 
+	@OneToMany(mappedBy = "issue", fetch = FetchType.LAZY)
+	private List<IssueAssignee> issueAssignees = new ArrayList<>();
+
```
해당부분은 리뷰주신 내용을 반영하여 양방향 매핑을 해주었습니다. 
이슈가 조회 될 때 IssueAssignee가 항상 필요한 부분이여서 양뱡향으로 매핑하였을 때 DB조회가 훨씬 편리함을 느꼈습니다. 

그리고 조금 더 정리해봐야겠지만 
OneToMany 일 때 Lazy 조회는 N+1문제가 발생하지 않음을 알게 되었습니다. 
하나의 조건에서 여러 개를 조회하는 것이니 당연히 N번 쿼리가 발생하지 않는다는 것을 알았습니다 

책으로 볼때는 당연한 지식이 JPA를 처음 사용하고 적용하다보니 구현할 때와는 다른 것을 느끼고 있습니다.

Dae-Hwa replied:
제가 착각하고 있는 것일 수도 있는데, 제가 생각했을때는 여기도 N+1이 발생해야 할 것 같아요. 그런데 조회가 한 번만 되네요..?
이건 저도 좀 알아보고 말씀드리겠습니다. 
https://jojoldu.tistory.com/165 N+1은 여기에 예제소스도 같이 올려놓으시니까 돌려보시면서 공부하시면 도움 될것 같아요

mybloom replied:
프레디~
공유해주신 향로님의 글을 다시 한번 정독하면서 연관 글(https://jojoldu.tistory.com/457)을 읽어보았는데요. 
해당 프로젝트에서 N+1이 발생하지 않은 이유를 발견하였습니다. 
코드의 해당 부분은 `이슈상세조회`에 해당하는 부분이라서 조회된 부모가 1개인 상황입니다. 
그래서 `@OneToMany`의 issueAssignees를 조회하는 쿼리가 1번이 되네요. 

만약에 목록조회등 부모, 여기서는 Issue가 여러 개 조회되었다면 연관된 IssueAssignees를 lazy 조회할 때 N+1의 문제가 발생했을 것입니다. 

이렇게 "상세 조회"처럼 부모가 1개일때는 eager 조회를 하지 않고, 현재의 방식대로 사용되도 될 것 같은 생각인데요. 
프레디의 생각이 궁금합니다.

Dae-Hwa replied:
아 그렇네요 부끄러워라 ㅎㅎㅎㅎ
단단히 착각하고 있었네요 감사합니다

eager는 특별한 경우가 아니면 지양하는게 좋다고 생각합니다. 그리고 OneToMany는 기본이 lazy에요 참고해주세요~

mybloom commented on BE/src/main/java/com/team09/issue_tracker/label/Label.java:
```diff
@@ -24,8 +30,15 @@ public class Label {
 
 	private String backgroundColor;
 
-	@Transient
-	private Long writerId;
+	private Long memberId;
```
테스트 데이터 셋팅이 어느 정도 되어서, 임시로 사용하였던 `@Transient`를 지우고 DB와 매칭되는 멤버변수를 사용하였습니다. 

Label에서는 Member 엔티티 자체가 필요없고, memberId로 누가 생성한 레이블인지만 표현하면 되어서 객체 매핑을 하지 않았습니다.

Dae-Hwa replied:
누가 생성한 레이블인지만 알면 된다는게 member는 추가적으로 조회할 필요가 없다는 뜻일까요?
만약 그게 아니라면 Label도 member를 알면 Label과 관련된 member를 조회해올 수 있지 않을까요?

mybloom replied:
현재는 Label을 조회할 때  memberId만 조회되는 상황인데요. 현재까지에서는 Member관련해서 다른 정보가 필요없어서 이렇게 해두었습니다.
Label과 관련된 member의 다른 정보(userId나 이름등)가 필요하다면 Label이 Member를 알아야 하겠네요.

mybloom committed:
refactor: LabelService boolean 리턴 표현 변경 (bugpigg/issue-tracker#17)

mybloom commented on BE/src/main/java/com/team09/issue_tracker/label/LabelRepository.java:
```diff
@@ -0,0 +1,15 @@
+public interface LabelRepository extends JpaRepository<Label, Long> {
+
+	//하나라도 존재하면 true
+	boolean existsByIdInAndMemberId(Set<Long> ids, Long memberId);
+	boolean existsByIdAndMemberId(Long id, Long memberId);
+
+	long countByIdInAndMemberId(List<Long> ids, Long memberId);
```
여러 쿼리메서드를 사용해보았습니다. 
exists는 하나라도 존재하면 true라서 , count함수를 사용하여 인자로 받은 List의 갯수와 count의 갯수가 동일한지 
validate 서비스로직에서 판별하였습니다.

Dae-Hwa replied:
실제 사용하지 않는 것은 포함시키지 않는게 좋습니다!

mybloom committed:
refactor: MilestoneService boolean 리턴 표현 변경 (bugpigg/issue-tracker#17)

mybloom commented on BE/src/test/java/com/team09/issue_tracker/issue/IssueCreateRepositoryTest.java:
```diff
@@ -24,24 +26,29 @@ public class IssueCreateRepositoryTest {
-			.build();
-		boolean isOpened = true;
-		Issue issue = Issue.fromForMandatory(issueSaveRequestDto, isOpened, MEMBER_ID);
+	/** TODO : repository 테스트인데, 일부 서비스 테스트 요소가 들어가 있는 것 같다.
+	 		 왜냐면, request parameter를 엔티티로 변환하는 과정이 들어가 있는 것 같다.
+	 **/
+//	@DisplayName("필수 요소만 가진 이슈")
+//	@Test
+//	void saveOnlyMandatory() {
```
테스트코드는 우선순위에 따라 리팩토링 해보도록 하겠습니다.

시도해보고 싶은 테스트코드는 저번 프로젝트에 주로 했던 repository 테스트코드와 
하게 된다면 이번에 처음 하게 될 Service 레이어 테스트코드를 생각하고 있습니다.

Dae-Hwa replied:
VCS를 사용하니 동작하지 않는 코드는 되도록이면 지워주는게 좋을 것 같습니다

mybloom replied:
사용하지 않는 코드는 삭제해서 올리겠습니다.

mybloom committed:
fix: @Transactional 애노테이션 중복 - @DataJpaTest 존재 (bugpigg/issue-tracker#17)

bugpigg commented on BE/src/main/java/com/team09/issue_tracker/login/filter/LoginValidateFilter.java:
```diff
@@ -0,0 +1,62 @@
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+		FilterChain filterChain) throws ServletException, IOException {
+
+		String requestUrl = request.getRequestURI();
+		if (isValidRequiredUrl(requestUrl)) {
+			log.info("로그인 인증 유무 확인 로직 실행 ---> {}", requestUrl);
+			String accessToken = getAccessToken(request);
+			if (!tokenProvider.validateToken(accessToken)) {
+				throw new RuntimeException("유효하지 않은 JWT 토큰입니다!!");
+			}
+			addMemberIdToResponseHeader(accessToken, response);
+		}
```
로그인이 필요한 URL에 접속시, 요청 헤더에서 `Authorization` 키의 밸류값인 `accessToken` 을 가져옵니다.
검증 과정을 거치고  `accessToken` 에서 가져온 `userId` 값을 이용해 `id` 를 찾아 응답 헤더에 추가합니다.

사실 요청 헤더에 추가하는게 의미적으로 맞다고 생각했지만, 구현상 어려움이 있어 우선 응답헤더에 추가하는 방식으로 구현하였습니다

Dae-Hwa replied:
응답에 userId가 포함돼도 상관 없는건가요?
내부에서 사용하는 용도면 다른 곳에 보관해도 상관없지 않을까요?

Dae-Hwa replied:
그리고 사용자를 찾는 작업은 인터셉터를 이용하는게 더 좋을 것 같습니다!

bugpigg replied:
오 넵 말씀 주신대로,  인터셉터와 필터에 대해 공부해보고 수정해보겠습니다!
다른 곳에 보관하는 방식도 찾아봐야 겠습니다!

Dae-Hwa replied:
그냥 동일한 리퀘스트 안에서만 유지돼도 상관없다면 일단은 RequestAttribute  같은데 넣어보고 이후에 생각해봐도 좋을 것 같아요

bugpigg committed:
fix: MemberType 이넘 클래스에 스트링 기반 이넘 객체 반환 기능 추가(#31)

- MemberType 이넘 클래스에 getEnum() 함수를 추가하여, 스트링 기반 값에 따라 매핑되는 이넘 객체를 반환합니다.
- 기존 이넘의 기본 스태틱 함수인 valueOf()를 사용하던 OauthUserProfile 클래스를, 새로 추가한 getEnum() 함수를 활용하도록 수정하였습니다.

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/exception/BusinessException.java:
```diff
@@ -0,0 +1,10 @@
+
+public abstract class BusinessException extends RuntimeException {
+
+	protected abstract HttpStatus getHttpStatus();
+
+	protected abstract String getBodyMessage();
```
이런 방식도 괜찮은 것 같아요 👍

Dae-Hwa replied:
https://codesquad-members.slack.com/archives/C01JLFFQA9E/p1655728248313449
얼마 전에 이런 글도 올라왔었는데 여기서 얘기하는 것 처럼 HTTP Status와 함께 예외 상황에 대한 코드를 함께 정의해서 응답에 포함시켜주면 클라이언트가 복잡한 로직을 수행할때 도움을 줄 수도 있어요 ㅎㅎ 이건 참고만 해주세요

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueController.java:
```diff
@@ -76,4 +81,22 @@ public ResponseEntity<CommonResponseDto> updateAllState(@RequestParam final Bool
 		return ResponseEntity.ok(new CommonResponseDto());
 	}
+
+	private void validateCreateRequestDto(IssueSaveRequestDto issueCreateRequestDto) {
```
객체에 대한 검증은 객체를 가져오면서 해줘야 어떤 객체를 어떤 상황에서 가져올 수 있는건지 명확하게 알 수 있을 것 같아요. 예를 들어서 마일스톤이 있는지 확인은 마일 스톤을 가져오면서 해주면 되겠죠

mybloom replied:
그럼 마일스톤을 가져오는 Service레이어에서 유효성 체크를 해주면 될까요?

mybloom replied:
객체 검증을 controller에서 하고, 사용은 service에서 해주었는데요.

검증을 service단에서 하도록 변경하겠습니다

Dae-Hwa replied:
컨트롤러에서 해야되는 검증은 컨트롤러의 책임에 맞게 이루어져야 한다고 생각해요
컨트롤러의 역할이 어떤건지 생각해보시면 좋을 것 같습니다!

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
-		return response;
+
+		//1. Milestone 생성
+		Milestone milestone = null;
+		if (milestoneId != null) {
+			milestone = Milestone.of(milestoneId);
+		}
```
Milestone에 대한 검증을 여기서 해주는게 좋을 것 같습니다. Milestone이 없으면 가져오지 못하는게 Issue 조회의 조건이니까요

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
+		}
+
+		//2. Issue 생성
+		IssueSaveServiceDto issueSaveServiceDto = IssueSaveServiceDto.builder()
```
서비스를 위한 DTO을 별도로 두고 싶다면 컨트롤러에서 변환해서 전달해주는게 좋겠네요
혹시 만든 이유가 서비스 레이어에서 컨트롤러 영역을 모르게 하고 싶어서가 아닌가요?

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
+			.isOpened(isOpened)
+			.memberId(memberId)
+			.build();
+		Issue issue = Issue.from(issueSaveServiceDto);
```
의존성의 방향을 중요한 부분으로 향하게 제어하고 싶다면, Issue를 도메인 객체라고 생각했을때 DTO를 모르게 하는게 좋을 것 같아요.
편의상 이렇게 만들 수도 있다고 생각하는데, IssueSaveServiceDto 를 만드신거 보고 커멘트 드립니다

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
+			issue.addIssueAssignee(savedIssueAssignees);
+		}
+
+		return savedIssue.toCommonResponse();
```
이것도 위에서 말씀드린거랑 마찬가지겠죠

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
 
 		//2. 할당자인지 확인
-		List<IssueAssignee> issueAssignees = issueAssigneeRepository.findByIssueId(issue.getId());
+		List<IssueAssignee> issueAssignees = issue.getIssueAssignees();
```
할당자인지는 어떻게 확인하는건가요?

mybloom replied:
Issue에서 Lazy 로딩으로 DB에서 조회해오고 있습니다.
이슈상세조회에만 Lazy로딩으로 할당자를 조회해오고 있습니다. 
부모가 하나(Issue)이면서 작성자가 아닐 경우에만 할당자 조회하는 로직이여서 eager조회할 필요가 없다고 생각해서입니다.

이슈 목록조회일 경우 할당자는 eager조회로 엔티티그래프로 가져오고 있어요

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -36,32 +42,80 @@ public CommonResponseDto create(IssueSaveRequestDto issueCreateRequestDto, Long
 
 		if (issue.isWriter(memberId)) {
 			isEditable = true;
+			return issue.toDetailResponse(isEditable);
```
이렇게 하기로 합의가 됐을거라 생각하고 지난번에는 말씀 안드렸던 것 같은데, 개인적으로 isEditable 같은 경우는 프런트에서 판단할 수도 있다고 생각해요
작성자인지 확인해야하는 곳은 수정하는 부분에서 확인이 돼야하는게 아닌가 하는 생각도 드네요

그리고 여러가지 상태가 올 수 있다면 `IssueDetailResponse.createEditable(issue)` 이런 식으로 할 수도 있을 것 같네요

mybloom replied:
상세화면에 "더보기" 버튼이 있는데요. 더보기 버튼 기능이 수정하는 화면을 보여주게 되어서 이슈상세 조회에 isEditable을 넣게 되었어요
생각해보니 클라이언트에서 해줘도 되겠네요. 클라와 이야기해보겠습니다.

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/dto/IssueSaveServiceDto.java:
```diff
@@ -0,0 +1,25 @@
+
+	private String title;
+	private String content;
+	private Milestone milestone;
+	private List<IssueLabel> issueLabels;
```
위에서 말씀드린거랑 연결시켜서 생각해보면 여기서도 id만 알고있으면 되지 않을까 싶네요

mybloom replied:
Service Dto에 대해서 잘 모르고 사용한 것 같습니다. 
프레디 리뷰 덕분에 Serivce단에서 Controller영역을 모르게 하는 용도인 것을 알았습니다.

제가 ServiceDto를 만들게 된 이유는 
엔티티를 생성하기 편하게 하기 위해서 만들었는데요. 메서드에 파라미터가 주렁주렁 달린 것이 마음에 들지 않아 만들게 되었습니다.
이름만 서비스dto이고 엔티티dto였네요.... 
해당 dto는 다음 pr때 없애도록 하겠습니다.

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/label/LabelService.java:
```diff
@@ -0,0 +1,18 @@
+
+	private final LabelRepository labelRepository;
+
+	public boolean isMyLabels(List<Long> labelIds, Long memberId) {
```
검증하는 이유가 어떤건지 궁금합니다

mybloom replied:
이슈 저장시 body에서 넘어온 id값들이 
DB에 존재하는 값이고, 이슈 작성자가 생성한 LabelId가 맞는지 확인하기 위함이었는데요.

해당 유효성확인이 불필요한 과정인지 궁금합니다.
검증하게 된 배경은 클라이언트에서 넘어 온 값들은 저장 전에 다 확인이 필요할 것 같은 생각에서 하게 되었습니다.

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/login/service/LoginService.java:
```diff
@@ -21,20 +27,29 @@ public class LoginService {
+	@Transactional
 	public LoginResponseDto login(String providerName, String code) {
-
+		// providerName에 해당하는 provider의 속성값을 제공하는 OauthProvider 객체
```
주석이 지나치게 많은 것 같아요
함수나 변수로 최대한 표현하려 해보시고 특이하거나 직관적이지 않은 동작에 대해서 주석이 있으면 좋을 것 같습니다.
그걸로 충분하지 않은 것 같다면 하나의 메소드에서 너무 많은 일을 하려 하는게 아닌지 생각해보시면 좋을 것 같습니다

bugpigg replied:
`LoginService` 클래스의 `login()` 메서드가 너무 많은 일을 한꺼번에 처리하다보니,
저도 코드를 읽는데 거북한 부분이 있어 주석을 많이 달게 되었습니다 ㅠㅠ 😂

말씀주신대로 메소드를 나누는 방식으로 수정해보겠습니다!!

Dae-Hwa replied:
넵 메소드를 나누다보면 자연스럽게 알맞은 클래스로 옮길 수 있게 될 것 같아요

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/milestone/MilestoneService.java:
```diff
@@ -0,0 +1,21 @@
+	@Transactional(readOnly = true)
+	public boolean isMyMilestone(Long milestoneId, Long memberId) {
+		Member member = Member.of(memberId);
+		long countOfMyMilestone = milestoneRepository.countByIdAndWriter(milestoneId, member);
```
exists로는 안 되나요?

mybloom replied:
exists로 하게 되면, 검증 파라미터가 여러 개 일 때 한개만 존재해도 true를 반환해서 모든 파라미터가 db에 있는 값인지 확인하면 count 갯수로 비교해야 했습니다

Dae-Hwa commented on BE/src/test/java/com/team09/issue_tracker/issue/studyofjpa/QueryMethodTest.java:
```diff
@@ -0,0 +1,107 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.transaction.annotation.Transactional;
+
+@DisplayName("메서드 쿼리 학습테스트")
```
마찬가지로 되도록이면 다른 브랜치에서 학습해보시면 더 좋을 것 같아요

Dae-Hwa commented on BE/src/test/java/com/team09/issue_tracker/issue/studyofjpa/QueryMethodTest.java:
```diff
@@ -0,0 +1,107 @@
+		Milestone milestone = Milestone.of(1L);
+
+		//when
+		Stream<Issue> findIssue = issueRepository.findByMilestone(milestone);
```
이건 사용하고 나서 detach해주시던가 닫아주셔야 합니다!

Dae-Hwa reviewed (approved):
리뷰가 늦어 죄송합니다

로직리뷰는 어떤 부분을 말씀하시는 걸까요? 

필터는 인터셉터랑 뭐가다른지 꼭 학습해보셨으면 좋겠어요

그리고 제가 지난번에 `EntityGraph`를 놓쳤는데 이 것과 `Stream`을 반환타입으로 갖는 레포지토리는 꼭 학습하고 사용해주세요!
